src/Microsoft.DotNet.Build.VstsBuildsApi/VstsDefinitionClient.cs

```diff
@@ -0,0 +1,178 @@
+        }
+
+        /// <summary>
+        /// Load a local release definition, combine the uniqueIdentifier with the definition name,
+        /// check VSTS for a matching definition name, if present, update that definition with the
+        /// local release definition, otherwise create a new definition.
+        /// </summary>
+        /// <param name="stream">Stream to a VSTS release definition</param>
+        /// <returns>Created or updated release definition id</returns>
+        public async Task<string> CreateOrUpdateReleaseDefinitionAsync(Stream stream)
+        {
+            JObject definition = GetParsedJson(stream);
+            var client = new VstsReleaseHttpClient(definition, _credentials);
+            return await CreateOrUpdateDefinitionAsync(client, definition).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Load a local definition, combine the uniqueIdentifier with the definition name, 
+        /// check VSTS for a matching definition name, if present, update that definition with the
+        /// local definition, otherwise create a new definition.
+        /// </summary>
+        /// <param name="client">Definition client that interacts with VSTS API and understands
+        /// definition JSON content.</param>
+        /// <param name="definition">VSTS definition JSON object</param>
+        /// <returns>Created or updated definition id</returns>
+        private async Task<string> CreateOrUpdateDefinitionAsync(
+            IVstsDefinitionHttpClient client,
+            JObject definition)
+        {
+            var key = _collectionIdentifier + "_" + definition["name"];
+            definition["name"] = key;
+            IReadOnlyList<JObject> vstsDefinitions = await client.VstsRetrieveDefinitionsListByNameAndPathAsync(definition).ConfigureAwait(false);
+
+            if (vstsDefinitions.Count == 0)
+            {
+                /* Create */
+
+                /* Remove definition instance identifiable information */
+                RemoveIdentifiableInformation(client, definition);
+
+                JObject vstsDefinition = await client.VstsCreateDefinitionAsync(definition).ConfigureAwait(false);
+                return vstsDefinition["id"].ToString();
+            }
+            else if (vstsDefinitions.Count == 1)
+            {
+                JObject vstsDefinition = await client.VstsRetrieveDefinitionByIdAsync(vstsDefinitions[0]).ConfigureAwait(false);
+
+                /* Update */
+                if (!IsDefinitionContentSubsetEquivalent(client, definition, vstsDefinition))
+                {
+                    CopyIdentifiableInformation(client, vstsDefinition, definition);
+                    vstsDefinition = await client.VstsUpdateDefinitionAsync(definition).ConfigureAwait(false);
+                }
+                return vstsDefinition["id"].ToString();
+            }
+            else
+            {
+                throw new Exception(string.Format(
```

Comment:
Kind of a weird one because it depends on what's on the server more than the state of the client itself (as `InvalidOperationException` docs suggest). I don't know of anything for this situation, so IOE sounds fine to me.